### PR TITLE
Add check collections to OkComputer

### PR DESCRIPTION
**Pull request by newzac**
Description and Impact
----------------------
> What impact does this change have for production?

Adds an API for having many check collections.

### CODE REVIEW

File: lib/ok_computer/check_collection.rb [hunk 1]
```diff
@@ -1,32 +1,80 @@
 module OkComputer
   class CheckCollection
-    attr_accessor :registry
+    attr_accessor :collection, :registrant_name, :display
 
     # Public: Initialize a new CheckCollection
     #
-    # registry - a Hash of checks, with keys being unique names and values
+    # collection - a Hash of checks, with keys being unique names and values
```
Comment (pbyrne) [inline lib/ok_computer/check_collection.rb]: Looks like `collection` isn't an argument we can actually pass in; and the new `display` option isn't documented.

File: lib/ok_computer/registry.rb [hunk 2]
```diff
@@ -9,43 +9,55 @@ class Registry
     # check_name - The name of the check to retrieve
     #
     # Returns the registered check or raises Registry::CheckNotFound
-    def self.fetch(check_name)
-      registry.fetch(check_name)
-    rescue KeyError
-      raise CheckNotFound, "No matching check"
+    def self.fetch(name)
+      check = default_collection.fetch(name)
```
Comment (pbyrne) [inline lib/ok_computer/registry.rb]: What do you think of this one-liner for this method?

```ruby
default_collection.fetch(name) || raise(CheckNotFound, "No matching check")
```

I think collapsing it a bit makes it clearer what's going on.

Comment (pbyrne) [inline lib/ok_computer/registry.rb]: Or at the very least collapse the next two lines into

```ruby
check || raise(CheckNotFound, "No matching check")
```

Comment (pbyrne) [inline lib/ok_computer/registry.rb]: Last thing… if `CheckCollection#fetch` mirrored `Hash#fetch`'s behavior of raising a `KeyError` when given a non-matching key, this method wouldn't need to change at all.

File: lib/ok_computer/check_collection.rb [hunk 3]
```diff
@@ -1,32 +1,80 @@
...
     end
 
-    # Public: Run the registry's checks
+    # Public: Run the collection's checks
     def run
       OkComputer.check_in_parallel ? check_in_parallel : check_in_sequence
     end
 
+    # Public: Returns a check or collection if it's in the check collection
+    def fetch(name, default=nil)
```
Comment (pbyrne) [inline lib/ok_computer/check_collection.rb]: Mirroring here my comment in `Registry#fetch` that it probably makes sense to mirror `Hash#fetch` in how it handles a key which doesn't exist: raising `KeyError`. If we do that, we wouldn't want to alias `:[]` to fetch; it'd become something like this instead:

```
def [](key)
  fetch(key)
rescue KeyError
end
```

File: lib/ok_computer/registry.rb [hunk 4]
```diff
@@ -9,43 +9,55 @@ class Registry
...
-    end
-
-    # Private: The list of registered checks, keyed by their unique names
-    #
-    # Returns a Hash
-    def self.registry
-      @registry ||= {}
+    # collection_name - The name of the check collection the check should be deregistered from
+    def self.deregister(check_name, collection_name=nil)
+      collection = collection_name ? default_collection.fetch(collection_name) : default_collection
```
Comment (pbyrne) [inline lib/ok_computer/registry.rb]: This duplicated logic with `#register` above seems to deserve its own named method (along with the `raise` in the next line).

File: spec/ok_computer/check_collection_spec.rb [hunk 5]
```diff
@@ -30,15 +35,76 @@ module OkComputer
...
     context "#checks" do
       it "returns the checks from its registry" do
-        subject.checks.should == registry.values
+        expect(subject.checks).to eq(registry.values)
+      end
+    end
+
+    context "#register" do
+      it "registers a check" do
+        subject.register(:foo, foocheck)
```
Comment (pbyrne) [inline spec/ok_computer/check_collection_spec.rb]: If we leave the `before` block above, I don't think this is checking anything. `foo` is already registered.

File: spec/ok_computer/check_collection_spec.rb [hunk 6]
```diff
@@ -6,11 +6,16 @@ module OkComputer
     let(:barcheck) { double(:check) }
     let(:registry) { {foo: foocheck, bar: barcheck} }
 
-    subject { CheckCollection.new registry }
+    subject { CheckCollection.new("foo collection name") }
 
-    context ".new registry" do
-      it "remembers the registry of checks given to it" do
-        subject.registry.should == registry
+    before do
```
Comment (pbyrne) [inline spec/ok_computer/check_collection_spec.rb]: In general, I find myself not a big fan of setting up state in a `before` block up at the outer context. Looking at tests down at the bottom of the file, this setup isn't visible and could be confusing to people who don't realize it's happening.

I'd rather perform this setup in any of the tests which need it, right in the test (or in a `before` block limited to just that `context` if it's burdensome to duplicate). A little bit of incidental duplication in tests doesn't bother me overmuch.

File: spec/ok_computer/check_collection_spec.rb [hunk 7]
```diff
@@ -30,15 +35,76 @@ module OkComputer
...
       end
     end
 
     context "#to_text" do
+      let(:foocheck) { Check.new }
+      let(:barcheck) { Check.new }
+
+      subject { CheckCollection.new("foo collection name") }
+
+      before do
```
Comment (pbyrne) [inline spec/ok_computer/check_collection_spec.rb]: To my point above about the `before` block at the top level of the test file, this isn't necessary here with that block in place above; these checks are already registered with every test already as things are now.

(That said, we should leave this here if we remove that other `before` block as I suggest.)

File: spec/ok_computer/check_collection_spec.rb [hunk 8]
```diff
@@ -30,15 +35,76 @@ module OkComputer
...
+        subject.register(:foo, foocheck)
+        subject.register(:bar, barcheck)
+      end
```
Comment (pbyrne) [inline spec/ok_computer/check_collection_spec.rb]: Can I be super not-picky and ask for a blank line between this `before` block and the test below it?

File: spec/ok_computer/registry_spec.rb [hunk 9]
```diff
@@ -2,38 +2,18 @@
...
-        Registry.instance_variable_set(:@registry, some_hash)
-        Registry.registry.should == some_hash
-      end
-
-      it "defaults to an empty hash if not set" do
-        Registry.instance_variable_set(:@registry, nil)
-        Registry.registry.should == {}
-      end
+    let(:check_object) { double(:first_checker, :registrant_name= => nil) }
+    let(:collection) { CheckCollection.new('foo collection') }
```
Comment (pbyrne) [inline spec/ok_computer/registry_spec.rb]: Similar to above; could we get a blank line after these`let`s?

File: spec/ok_computer/registry_spec.rb [hunk 10]
```diff
@@ -2,38 +2,18 @@
...
+    before do
```
Comment (pbyrne) [inline spec/ok_computer/registry_spec.rb]: Like the feedback on `before` above, I'd rather this be moved to the tests which need it rather than be performed for every test.

File: lib/ok_computer/registry.rb (hunk 2 above)
Comment (pbyrne) [inline lib/ok_computer/registry.rb]: > we were keeping the same behavior 

Understood. My concerns are more with the implementation than the behavior. I think by the end of my stream of consciousness, I landed on wishing the new `CheckCollection#fetch` method raised a `KeyError` meaning this method wouldn't have to change in the first place.

Barring that, collapsing down to a one-liner feels clearer than the current implementation (while providing the same behavior).

**Comment by anfleene** [comment]
@pbyrne I think we've addressed all of your comments. Want to give it a final look over?

File: lib/ok_computer/check_collection.rb [hunk 11]
```diff
@@ -1,32 +1,88 @@
...
+    attr_accessor :collection, :registrant_name, :display
 
     # Public: Initialize a new CheckCollection
     #
-    # registry - a Hash of checks, with keys being unique names and values
-    #   being Check instances
-    def initialize(registry={})
-      self.registry = registry
+    # display - the display name for the Check Collection
+    def initialize(display)
```
Comment (pbyrne) [inline lib/ok_computer/check_collection.rb]: Does this constitute a breaking change requiring a version bump? In theory, nobody's called this, since it's not part of the documented API. But it's a public method whose interface is changing.

Alternately, could we reasonably tweak this such that it _doesn't_ make an API-incompatible change?

Comment (anfleene) [inline lib/ok_computer/check_collection.rb]: hmm I feel like it's fine since it's not part of the public API. Not sure how easily we could maintain compatibility.

Comment (pbyrne) [inline lib/ok_computer/check_collection.rb]: I trust your judgement here.
